File: Top Interview Question - Easy Collection/array_threeSum.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int> > threeSum(vector<int> &num){
    vector<vector<int>> res;

    sort(num.begin(), num.end());

    for(int i = 0; i < num.size(); i++) {
        int target = -num[i];
        int begin = i+1;
        int end = num.size();

        while(begin < end) {
            int sum = num[begin] + num[end];

            if(sum < target)
                begin++;
            
            else if(sum > target) 
                end--;
            
            else {
                vector<int> triple(3,0);
                triple[0] = num[i];
                triple[1] = num[begin];
                triple[2] = num[end];
                res.push_back(triple);

                while(begin < end && num[begin] == triple[1]) begin++;
                while(begin > end && num[begin] == triple[2]) end--;
            }
        }
        while(i+1 < num.size() && num[i+1] == num[i])
            i++;
    }  
    return res;
}

int main() {
    vector<int> num = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> res = threeSum(num);
    for(int i = 0; i < res.size(); i++){
        cout<<"[";
        for(int j = 0; j < res[i].size(); j++){
            cout<<res[i][j] << " ";
        }
        cout<< "]";
    }
    cout <<"\n";
}

File: Leetcode - Python/array_threeSum.py
"""
Three Sum : https://leetcode.com/problems/3sum/
Author : Sharath Nagendra

Passes all leetcode test cases as on Oct 4, 2019
Runtime: 552 ms, faster than 90.28% of Python online submissions for 3Sum.
Memory Usage: 15 MB, less than 32.69% of Python online submissions for 3Sum.
"""
def threeSum(nums):
    res = []
    nums.sort()
    length = len(nums)
    for i in range(length-2):
        if nums[i] > 0: break
        if i > 0 and nums[i] == nums[i-1]: continue 

        l, r = i+1, length-1
        while l < r:
            total = nums[l] + nums[i] + nums[r]
            if total < 0:
                l += 1
            elif total > 0:
                r -= 1
            else:
                res.append(nums[i], nums[l], nums[r])
                while l < r and nums[l] == nums[l+1]:
                    l += 1
                while l < r and nums[r] == nums[r-1]:
                    r -= 1
                l += 1
                r -= 1
    return res

nums = [-1, 0, 1, 2, -1, -4]
print(threeSum(nums))